### PR TITLE
Adding wrapping-layouts guide

### PR DESCRIPTION
**Pull request by alpertunga-bile**
## Description

Hello, I wrote a guide about wrapping the layout classes. I followed the contribution rules from the [guides/CONTRIBUTING.md](https://github.com/gradio-app/gradio/blob/main/guides/CONTRIBUTING.md) file.

- This pull request includes ```wrapping-layouts.md``` file which is located under the guides folder.
- The guide contains these features as stated in the ```CONTRIBUTING.md``` file:
   - An introduction part
   - Subheadings
   - Code snippets
   - Huggingface Space example
   - Tags at the top of the document
- Here is the introduction part of the guide:

### Introduction
Gradio features [blocks](https://www.gradio.app/docs/blocks) to easily layout applications. To use this feature, you need to stack or nest layout components and create a hierarchy with them. This implementation isn't difficult to implement and maintain for small projects, but after the project gets complex, this component hierarchy becomes difficult to maintain and reusing layouts becomes difficult.

In this guide, we are going to explore how we can wrap the layout classes to create more maintainable and easy-to-read applications without sacrificing flexibility.

EDIT : I relocated the ```wrapping-layouts.md``` file to ```09_other-tutorials``` folder to fix the building error.

### CODE REVIEW

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+        self.global_children_dict = {}
+        self.renderables = []
+```
+
+- Here are the variables' definitions:
+
+|       Variable       |              Type              | Definition                                                                                                                                        |
+| :------------------: | :----------------------------: | :------------------------------------------------------------------------------------------------------------------------------------------------ |
+|     main_layout      |      gradio.blocks.Block       | Stores the main layout from the gradio package of this class. For example, storing ```gradio.layouts.Row``` layout for the ```RowLayout``` class. |
+|         name         |              str               | Name of the class. Using to differentiate from other layouts and for debug purposes.                                                              |
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:39]:
```suggestion
|         name         |              str               | Name of the class. Used to differentiate from other layouts and for debugging purposes.                                                              |
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+| global_children_dict | Dict[str, gradio.blocks.Block] | Stores the children components with given name.                                                                                                   |
+|     renderables      |              list              | Stores the renderable elements such as components and layouts.                                                                                    |
+
+---------------------------------------------
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:43]: Why are you using this line breaks everywhere? They don't look so good when they're rendered.

Comment (aliabd) [reply guides/09_other-tutorials/wrapping-layouts.md:43]: Maybe if you want to seperate the sections you can use numbers for the big headings and bullet points for the smaller ones?

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+
+```python
+    # other LayoutBase implementations
+
+    def add_layout(self, layout: LayoutBase) -> None:
+        self.renderables.append(layout)
+        self.global_children_dict.update(layout.global_children_dict)
+```
+
+- As you can see, the ```global_children_dict``` variable is updating the dictionary with the added new layout's ```global_children_dict``` variable. With this functionality, the parent class includes all the components that children layouts have. You can see in the ```_attach_event``` function why this is important.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:71]:
```suggestion
- As you can see, the ```global_children_dict``` variable is updating the dictionary with the added new layout's ```global_children_dict``` variable. With this functionality, the parent class includes all the components that children layouts have. You can see in the ```attach_event``` function why this is important.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+- Now the last function we have to implement in the ```LayoutBase``` class is the ```attach_event``` function. The function is left as not implemented because it is specific to the class, so each class has to implement its `attach_event` function.
+
+```python
+    # other LayoutBase implementations
+
+    def attach_event(self, block_dict: Dict[str, Block]) -> None:
+        raise NotImplementedError
+```
+
+- You can see what is the ```block_dict``` variable in the ```_attach_event``` function which is implemented in the ```Application``` class.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:171]:
```suggestion
- You can see what is the ```block_dict``` variable in the ```attach_event``` function which is implemented in the ```Application``` class.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+
+    def attach_event(self, block_dict: Dict[str, Block]) -> None:
+        self.left_textbox.change(
+            change_text,
+            inputs=self.left_textbox,
+            outputs=self.right_textbox,
+        )
+```
+- As you can see from the constructor of the ```RowExample``` class, we add the ```render=False``` argument when creating the Textbox variables. This is an important parameter to add because we are calling the render function ourselves, so we don't want it to be called in the constructor of the component.
+- [x] As a note; even if you don't assign False to the render variable, the program works and is structured correctly but it is a good practice I think.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:468]:
```suggestion
- As a note; even if you don't assign False to the render variable, the program works and is structured correctly but it is a good practice I think.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+
+- Let's write a simple example to see how we can use the classes that we have created.
+- In this example, we are going to use two tabs, one row and one column layouts.
+- Row and column layouts are going to have two textboxes each.
+- Row's and column's second textboxes are going to be attached to the row's first textbox's value.
+- Here is the example we are going to write:
+
+<iframe
+	src="https://huggingface.co/spaces/WoWoWoWololo/wrapping-layouts"
+></iframe>
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:438]:
```suggestion
<gradio-app space="WoWoWoWololo/wrapping-layouts"></gradio-app>
```

**Review by aliabd — COMMENTED** [review]
Thanks @alpertunga-bile for writing this guide!

You can see a preview of the guide here by the way: 

https://gradio-34zk5e103-hugging-face.vercel.app/main/guides/wrapping-layouts 

Or just click on the website preview link from the gradio-pr-bot comment. 

I left some comments and suggestions, especially replacing the iframe with a gradio embed. 

If there's a way you can simplify the guide, it would be better. It's a little bit long and hard to follow. Maybe we can embed more spaces to show what things look like at intermediate stages? or remove some portions that aren't necessary?

**Comment by aliabd** [comment]
@alpertunga-bile is this ready for another review? Let me know if you're stuck on anything.

**Comment by alpertunga-bile** [comment]
Hello @aliabd. Sorry for the delay. I'm currently trying to rewrite the guide so it's easier to follow and read. I'll let you know when it's ready for review.

**Comment by alpertunga-bile** [comment]
Hello @aliabd. I made these changes:
  - Most of the bullet point elements have been removed
  - Gradio application integration was fixed
  - Most of the explanation was transferred to [the example code](https://huggingface.co/spaces/WoWoWoWololo/wrapping-layouts/blob/main/app.py)
  - The guide was shortened

The guide is ready for another review.

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+```python
+# self.main_layout -> Row()
+with self.main_layout:
+    left_textbox.render()
+    right_textbox.render()
+```
+
+So with calling the components' render functions under the ```with``` syntax, we are simulating the default implementation, actually.
+
+If we understand the below ```with``` syntax, let's move on the upper ```with``` syntax. For this, let's expand our example with ```Tab``` component:
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:80]:
```suggestion
So now let's consider two nested ```with```s to see how the outer one works. For this, let's expand our example with the ```Tab``` component:
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+```python
+with Tab():
+    with Row():
+        first_textbox = Textbox(value="first_textbox")
+        second_textbox = Textbox(value="second_textbox")
+```
+
+Pay attention to the Row and Tab components this time. We have created the Textbox variables above and added them to Row with ```with``` syntax. Now we need to add the Row component to the Tab component. You can see that Row component is created with default parameters, so its render parameter is true, that's why the render function is going to be executed under the Tab component's ```with``` syntax.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:89]:
```suggestion
Pay attention to the Row and Tab components this time. We have created the Textbox variables above and added them to Row with the ```with``` syntax. Now we need to add the Row component to the Tab component. You can see that the Row component is created with default parameters, so its render parameter is true, that's why the render function is going to be executed under the Tab component's ```with``` syntax.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        second_textbox.render()
+
+    row_main_layout.render()
+
+tab_main_layout.render()
+```
+
+The default implementation and our implementation are the same, but we are using the render function ourselves. So it requires a little work.
+
+Now, let's look at the ```attach_event``` function.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:108]:
```suggestion
Now, let's take a look at the ```attach_event``` function.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    The function is left as not implemented because it is specific to the class, so each class has to implement its `attach_event` function.
+
+```python
+    # other LayoutBase implementations
+
+    def attach_event(self, block_dict: Dict[str, Block]) -> None:
+        raise NotImplementedError
+```
+
+You can see what is the ```block_dict``` variable in the ```Application``` class's ```attach_event``` function.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:121]:
```suggestion
Check out the ```block_dict``` variable in the ```Application``` class's ```attach_event``` function.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        with self.app:
+            for child in self.children:
+                child.render()
+
+        self.app.render()
+```
+
+From the explanation of the ```LayoutBase``` class's ```render``` function, we can understand the ```child.render``` part.
+
+So let's look at the below part, why we are calling the ```app``` variable's ```render``` function? We are calling this render function because if we look at the implementation of this function in the ```gradio.blocks.Blocks``` class, we can see that it is adding the components and event functions into the root component. With another saying, it is creating and structuring the gradio application. So it is important to call this function.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:140]:
```suggestion
So let's look at the bottom part, why are we calling the ```app``` variable's ```render``` function? It's important to call this function because if we look at the implementation in the ```gradio.blocks.Blocks``` class, we can see that it is adding the components and event functions into the root component. To put it another way, it is creating and structuring the gradio application. 
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+            for child in self.children:
+                try:
+                    child.attach_event(block_dict=block_dict)
+                except NotImplementedError:
+                    print(f"{child.name}'s attach_event is not implemented")
+```
+
+You can see why the ```global_children_list``` is used in the ```LayoutBase``` class from the example code. With this, all the components in the application are gathered into one dictionary, so the component can access all the components with their names.
+
+```with``` syntax is used here again to attach events to components. If we look at the ```__exit__``` function in the ```gradio.blocks.Blocks``` class, we can see that it is calling the ```attach_load_events``` function which is used for setting event triggers to components. So we have to use the ```with``` syntax to trigger the ```__exit__``` function.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:165]:
```suggestion
The ```with``` syntax is used here again to attach events to components. If we look at the ```__exit__``` function in the ```gradio.blocks.Blocks``` class, we can see that it is calling the ```attach_load_events``` function which is used for setting event triggers to components. So we have to use the ```with``` syntax to trigger the ```__exit__``` function.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+Yes, we can call ```attach_load_events``` function without using ```with``` syntax, but the function needs a ```Context.root_block```, and it is set in the ```__enter__``` function. So we used ```with``` syntax here rather than calling functions ourselves.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:167]:
```suggestion
Of course, we can call ```attach_load_events``` without using the ```with``` syntax, but the function needs a ```Context.root_block```, and it is set in the ```__enter__``` function. So we used the ```with``` syntax here rather than calling the function ourselves.
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+## Conclusion
+
+In this guide, we saw
+
+- How we can wrap the layouts
+- How components are rendered
+- How we can structure our application with wrapped layout classes
+
+Because the classes used in this guide are used for demonstration purposes, they may lack optimizations or useful functionality. But this is the subject of another guide.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:177]:
```suggestion
Because the classes used in this guide are used for demonstration purposes, they may still not be totally optimized or modular. But that would make the guide much longer! 
```

File: guides/09_other-tutorials/wrapping-layouts.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+I hope this guide helps you to gain another view to look at the layout classes and gives you an idea about how you can use them for your needs.
```
Comment (aliabd) [inline guides/09_other-tutorials/wrapping-layouts.md:179]:
```suggestion
I hope this guide helps you gain another view of the layout classes and gives you an idea about how you can use them for your needs. See the full implementation of our example [here](https://huggingface.co/spaces/WoWoWoWololo/wrapping-layouts/blob/main/app.py).
```

**Review by aliabd — COMMENTED** [review]
Thanks @alpertunga-bile, guide is so much better now! 

Left some small linguistic suggestions, but this should be ready to merge soon!

**Comment by alpertunga-bile** [comment]
Hello @aliabd, thanks for the review and suggestions. I added the suggestions you give and checked the preview website for the layout errors. The guide is ready for another review.

**Review by aliabd — APPROVED** [review]
Everything LGTM! 🎉 

Thanks @alpertunga-bile. Will merge now
